From 04860bce8280064390fc0190642597a05ea8c6d7 Mon Sep 17 00:00:00 2001
From: Joshua Wilkinson <joshua.wilkinson@morsemicro.com>
Date: Wed, 30 Apr 2025 22:03:53 +0000
Subject: [PATCH 6/7] Merged in SW-14454-add-DPP-GAS-req-retries (pull request
 #286)
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

SW-14454 Add GAS req retries to DPP-PB flow

After DPP-PB successfully completes DPP Authentication, the Enrollee sends a Configuration Request to the Configurator.

In the Wi-Fi EasyConnect Specification (v2.0/v3.0), the relevant section is 7.4.4.2 - Provisioning State, which describes the actions taken in this state:

If t expires and an ACK frame was received for the previously transmitted DPP Configuration Request frame, the Enrollee
generates a NoPeer event and transitions to the Terminated state. Otherwise, if t expires, the state machine checks the
value of Sent. If Sent is five or less, then Sent is incremented, t is set to 10 seconds, the DPP Configuration Request
frame is resent, and the state machine remains in the Provisioning state. If Sent is greater than five, the Enrollee
generates a NoPeer event and transitions to the Terminated state.

Add capability for wpa_supplicant, when acting as an enrollee to send this frame a total of 5 times if it is not successful.

Note that currently, the hostap implementation of the other timeouts of this kind “t is set to 10 seconds”, are not followed (and doing so would massively increase the association time of DPP-PB).

Approved-by: Andrew Pope
Approved-by: Nev Young
Approved-by: James Haggerty
---
 src/common/dpp.h                |  1 +
 wpa_supplicant/dpp_supplicant.c | 25 +++++++++++++++++++------
 2 files changed, 20 insertions(+), 6 deletions(-)

diff --git a/src/common/dpp.h b/src/common/dpp.h
index e29f20952..237525e9b 100644
--- a/src/common/dpp.h
+++ b/src/common/dpp.h
@@ -340,6 +340,7 @@ struct dpp_authentication {
 	int waiting_auth_conf;
 	int auth_req_ack;
 	unsigned int auth_resp_tries;
+	unsigned int conf_req_tries;
 	u8 allowed_roles;
 	int configurator;
 	int remove_on_tx_status;
diff --git a/wpa_supplicant/dpp_supplicant.c b/wpa_supplicant/dpp_supplicant.c
index ce4bc1f54..4be3ada37 100644
--- a/wpa_supplicant/dpp_supplicant.c
+++ b/wpa_supplicant/dpp_supplicant.c
@@ -1899,8 +1899,16 @@ static void wpas_dpp_gas_resp_cb(void *ctx, const u8 *addr, u8 dialog_token,
 	}
 	if (result != GAS_QUERY_SUCCESS ||
 	    !resp || status_code != WLAN_STATUS_SUCCESS) {
-		wpa_printf(MSG_DEBUG, "DPP: GAS query did not succeed");
-		goto fail;
+		wpa_printf(MSG_DEBUG, "DPP: GAS query did not succeed (try: %d)",
+			auth->conf_req_tries);
+
+		if (auth->conf_req_tries >= 5)
+			goto fail;
+
+		auth->conf_req_tries++;
+		wpas_dpp_start_gas_client(wpa_s);
+		wpa_printf(MSG_DEBUG, "DPP: Restarting GAS client to resend GAS request");
+		return;
 	}
 
 	wpa_hexdump_buf(MSG_DEBUG, "DPP: Configuration Response adv_proto",
@@ -2054,8 +2062,10 @@ static void wpas_dpp_start_gas_client(struct wpa_supplicant *wpa_s)
 	 * out new Config Requests for 60 seconds, so this gives an extra 60
 	 * second time after the last expected new Config Request for the
 	 * Configurator to determine what kind of configuration to provide. */
-	eloop_register_timeout(120, 0, wpas_dpp_gas_client_timeout,
-			       wpa_s, NULL);
+	if (!eloop_is_timeout_registered(wpas_dpp_gas_client_timeout, wpa_s, NULL)) {
+		eloop_register_timeout(120, 0, wpas_dpp_gas_client_timeout,
+				       wpa_s, NULL);
+	}
 
 	res = gas_query_req(wpa_s->gas, auth->peer_mac_addr, auth->curr_freq,
 			    1, 1, buf, wpas_dpp_gas_resp_cb, wpa_s);
@@ -2086,10 +2096,12 @@ static void wpas_dpp_auth_success(struct wpa_supplicant *wpa_s, int initiator)
 	}
 #endif /* CONFIG_TESTING_OPTIONS */
 
-	if (wpa_s->dpp_auth->configurator)
+	if (wpa_s->dpp_auth->configurator) {
 		wpas_dpp_start_gas_server(wpa_s);
-	else
+	} else {
+		wpa_s->dpp_auth->conf_req_tries = 1;
 		wpas_dpp_start_gas_client(wpa_s);
+	}
 }
 
 
@@ -2744,6 +2756,7 @@ wpas_dpp_rx_reconfig_auth_conf(struct wpa_supplicant *wpa_s, const u8 *src,
 	if (dpp_reconfig_auth_conf_rx(auth, hdr, buf, len) < 0)
 		return;
 
+	auth->conf_req_tries = 1;
 	wpas_dpp_start_gas_client(wpa_s);
 }
 
-- 
2.34.1

